src/napari_threedee/data_models/dipoles.py

```diff
@@ -0,0 +1,163 @@
+import os
+from typing import List, Tuple
+
+import napari
+import numpy as np
+import zarr
+from pydantic import BaseModel, validator
+
+from napari_threedee.annotators.base import N3dDataModel
+from napari_threedee.annotators.constants import N3D_METADATA_KEY, ANNOTATION_TYPE_KEY
+from napari_threedee.data_models.spline_sampler import SplineSampler
+from napari_threedee.annotators.dipoles.validation import validate_layer, validate_n3d_zarr
+
+from napari_threedee.annotators.dipoles.constants import (
+    DIPOLE_ANNOTATION_TYPE_KEY,
+    DIPOLE_DIRECTION_X_FEATURES_KEY,
+    DIPOLE_DIRECTION_Y_FEATURES_KEY,
+    DIPOLE_DIRECTION_Z_FEATURES_KEY,
+)
+
+
+class N3dDipole(BaseModel):
+    center: np.ndarray
+    direction: np.ndarray
+
+    class Config:
+        allow_mutation = False
+        arbitrary_types_allowed = True
+
+    @validator('center', 'direction', pre=True)
+    def ensure_float32_ndarray(cls, value):
+        return np.asarray(value, dtype=np.float32)
+
+
+class N3dDipoles(N3dDataModel):
+    data: List[N3dDipole]
+
+    @property
+    def centers(self) -> np.ndarray:
+        if not self.data:
```

Comment:
```suggestion
        if len(self.data) == 0:
```

minor nit: this is more explicit and doesn't rely on understanding python specifics to read